### DISCUSSION TITLE
tutorials/inputs/handling_quit_requests

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#87] [User-contributed notes]
# tutorials/inputs/handling_quit_requests

Quitting: Most platforms have the option to request the application to quit. On desktops, this is usually done with the "x" icon on the window title bar. On mobile devices, the app can quit at any ...

https://docs.godotengine.org/en/stable/tutorials/inputs/handling_quit_requests.html

**Comment by shenkes** [comment]
This did not quit out for me in 4.3. It seems like you should use get_tree().quit() now according to this issue:
https://github.com/godotengine/godot/issues/67379

FWIW, I didn't dive in too deep to double check things.

**Reply by shenkes** [reply]
Specifically
get_tree().root.propagate_notification(NOTIFICATION_WM_CLOSE_REQUEST)
did not quite out for me without any other changes.

**Reply by Calinou** [reply]
The note at the end of the page covers this:

> Sending this notification will inform all nodes about the program termination, but will not terminate the program itself unlike in 3.X. In order to achieve the previous behavior, [SceneTree.quit](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-quit) should be called after the notification.

**Comment by Diamond-Dan** [comment]
So OS.kill cannot run after a quit command has been given which results in not being able to close out programs run by Godot

**Comment by AccountNameIsTaken** [comment]
Is there a way to use this to kill browser tabs in web builds? Even if I send the close request notification before quitting, web builds seem to just hang after quit.

**Comment by 23elemental23** [comment]
how do I declare NOTIFICATION_WM_QUIT_REQUEST? Because its not declared by default?

**Reply by Calinou** [reply]
It is `NOTIFICATION_WM_CLOSE_REQUEST` in Godot 4. `NOTIFICATION_WM_QUIT_REQUEST` was the name in Godot 3.

**Comment by vektor451** [comment]
`get_tree().quit()` does not seem to be the default quit behaviour.
`get_tree().quit()` when called manually, will iterate through the tree, and call every Node's `_exit_tree()` function. 

At least on windows, when you close via Alt-F4 or the window's close button, it skips this step, and the tree isn't allowed to clean up before the application closes. With the example code provided to handle quit requests, and calling `get_tree.quit()` upon the noticiation being recieved, the tree gets cleaned up properly.

**Reply by Calinou** [reply]
I suggest opening an issue for this on the main Godot repository. It's something I noticed when using Movie Maker mode too (AVI files won't have a proper footer written to them when closing with the X button or <kbd>Alt + F4</kbd>).
